### PR TITLE
Damage Formula

### PR DESCRIPTION
**Pull request by Zintixx**
- I believe this is the correct damage formula here but if not very close?
- Fixed buffs resetting upon changing maps
- Refresh Stats if there is a change in buffs, equips, etc.

### CODE REVIEW

**Review by AngeloTadeucci — APPROVED** [review]

File: Maple2.Server.Core/Formulas/BonusAttack.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            JobCode.Archer => 1.143,
+            JobCode.HeavyGunner => 1.364,
+            JobCode.Thief => 1.151,
+            JobCode.Assassin => 1.114,
+            JobCode.RuneBlader => 1.259,
+            JobCode.Striker => 1.264,
+            JobCode.SoulBinder => 1.177,
+            _ => 1,
+        };
+    }
```
[the remark this answers is not in the document]
Comment (Zintixx) [reply Maple2.Server.Core/Formulas/BonusAttack.cs]: nothing unfortunately. i dont see it in the lua script either

File: Maple2.Server.Game/Util/DamageCalculator.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+        // Get hit rate
+        if (Damage.Evade(target.Buffs.GetResistance(BasicAttribute.Accuracy), target.Stats.Values[BasicAttribute.Evasion].Total, target.Stats.Values[BasicAttribute.Dexterity].Total,
+                caster.Stats.Values[BasicAttribute.Accuracy].Total, caster.Buffs.GetResistance(BasicAttribute.Evasion), caster.Stats.Values[BasicAttribute.Luck].Total)) {
+            return (DamageType.Miss, 0);
+        }
+
+        var damageType = DamageType.Normal;
+        (double minBonusAtkDamage, double maxBonusAtkDamage) = caster.Stats.GetBonusAttack(target.Buffs.GetResistance(BasicAttribute.BonusAtk), target.Buffs.GetResistance(BasicAttribute.MaxWeaponAtk));
+        double attackDamage = minBonusAtkDamage + (maxBonusAtkDamage - minBonusAtkDamage) * Random.Shared.NextDouble();
```
Comment (mettaursp) [inline Maple2.Server.Game/Util/DamageCalculator.cs:30]: I feel like instead if minBonus and maxBonus it should just be minAttack and maxAttack. Bonus damage is its own stat with its own implications, while the output of this function is just broader results of all attack calculations. Also might need a way to disable the randomness cause `isConstDotDamageValue` & `isConstDamageValue` both exist. I have not looked into exactly how these work, but they're something to keep in mind.

File: Maple2.Server.Game/Util/DamageCalculator.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+        // change the NPCNormalDamage to be changed depending on target?
+        double damageBonus = 1 + (caster.Stats.Values[SpecialAttribute.TotalDamage].Multiplier()) + (caster.Stats.Values[SpecialAttribute.NormalNpcDamage].Multiplier());
+        if (target is FieldNpc npc && npc.Value.IsBoss) {
+            damageBonus += caster.Stats.Values[SpecialAttribute.BossNpcDamage].Multiplier();
+        }
+
+        // Get elemental bonus
+        double elementBonus = 0;
+        switch (record?.SkillMetadata.Property.Element) {
```
Comment (mettaursp) [inline Maple2.Server.Game/Util/DamageCalculator.cs:40]: Dots can have elements too.

File: Maple2.Server.Game/Util/DamageCalculator.cs [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                break;
+            case RangeType.Range:
+                damageBonus += caster.Stats.Values[SpecialAttribute.RangedDamage].Multiplier();
+                break;
+        }
+
+        damageBonus += -target.Buffs.GetResistance(BasicAttribute.AttackSpeed) * caster.Stats.Values[BasicAttribute.AttackSpeed].Total;
+
+        // Check for crit and get crit damage
+        if (!dotDamage && record != null && record.AttackMetadata.CompulsionTypes.Contains(CompulsionType.Critical)) {
```
Comment (mettaursp) [inline Maple2.Server.Game/Util/DamageCalculator.cs:77]: Dots can crit based on their useGrade. Both dots and skill hits can be guaranteed crits from soaring focus tonic, sin's fatal strikes, and specific skills having compulsionType 2.

File: Maple2.Server.Game/Util/DamageCalculator.cs (hunk 2 above)
Comment (Zintixx) [reply Maple2.Server.Game/Util/DamageCalculator.cs:30]: I believe we are using `isConstDotDamageValue`. The damage calculation will never happen and would just use the HpValue defined. is that correct?

Comment (mettaursp) [reply Maple2.Server.Game/Util/DamageCalculator.cs:30]: Oh I'm actually not sure. Sounds right enough

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**